Sources/NIOEmbedded/AsyncEmbeddedEventLoop.swift

```diff
@@ -0,0 +1,357 @@
+        await self.advanceTime(to: self.now)
+    }
+
+    /// Runs the event loop and moves "time" forward by the given amount, running any scheduled
+    /// tasks that need to be run.
+    public func advanceTime(by increment: TimeAmount) async {
+        await self.advanceTime(to: self.now + increment)
+    }
+
+    /// Unwrap a future result from this event loop.
+    ///
+    /// This replaces `EventLoopFuture.get()` for use with `AsyncEmbeddedEventLoop`. This is necessary because attaching
+    /// a callback to an `EventLoopFuture` (which is what `EventLoopFuture.get` does) requires scheduling a work item onto
+    /// the event loop, and running that work item requires spinning the loop. This is a non-trivial dance to get right due
+    /// to timing issues, so we need to
+    public func awaitFuture<ResultType: Sendable>(_ future: EventLoopFuture<ResultType>) async throws -> ResultType {
+        // We need a task group to wait for the scheduled future result, because the future result callback
+        // can only complete when we actually run the loop, which we need to do in another Task.
+        return try await withThrowingTaskGroup(of: ResultType.self, returning: ResultType.self) { group in
+            group.addTask {
+                try await future.get()
+            }
+
+            group.addTask {
+                while true {
+                    await self.run()
+                    try Task.checkCancellation()
+                    try await Task.sleep(nanoseconds: 100_000)
+                }
+            }
+
+            do {
+                // This force-unwrap is safe: there are only two tasks, one will either return or hang, and the
+                // one that returns cannot be nil, only error. We then cancel the other Task, but we'll never
+                // ask for its result.
+                let result = try await group.next()!
+                group.cancelAll()
+                return result
+            } catch {
+                group.cancelAll()
+                throw error
+            }
+        }
+    }
+
+    /// Runs the event loop and moves "time" forward to the given point in time, running any scheduled
+    /// tasks that need to be run.
+    ///
+    /// - Note: If `deadline` is before the current time, the current time will not be advanced.
+    public func advanceTime(to deadline: NIODeadline) async {
+        await withCheckedContinuation { (continuation: CheckedContinuation<Void, Never>) in
+            self.queue.async {
+                let newTime = max(deadline, self.now)
+
+                while let nextTask = self.scheduledTasks.peek() {
+                    guard nextTask.readyTime <= newTime else {
+                        break
+                    }
+
+                    // Now we want to grab all tasks that are ready to execute at the same
+                    // time as the first.
+                    var tasks = Array<EmbeddedScheduledTask>()
```

Comment:
Minor: should we use a circular buffer and move it outside of the while-loop to avoid allocating an array for each distinct `readyTime`?